### DISCUSSION TITLE
Units for energy consumption by distance

### DISCUSSION THREAD

**Discussion by jschlyter** [home-assistant/architecture#1163] [Entity Models] [answered]
Home Assistant currently has no units for measuring energy consumption by distance (e.g., kWh/100km or mi/kWh). Energy consumption per time unit , e.g. kWh/h is usually written as kW so that is already covered.

Energy consumed by electric cars is typically measured in kWh/100km in Europe and in mi/kWh in US/Canada. This is similar to fuel consumption with l/10km in Europe and miles/gallon in US/Canada. I have a [draft PR](https://github.com/home-assistant/core/pull/136933) introducing _energy distance_ units supporting all these units and conversion between them.

## Proposal

Add an `energy_distance` sensor device class with support for unit conversion between metric and US customary energy consumption units by distance, including their inverses (Europe normally uses energy per distance whereas US/UK uses distance per energy). The following units should be supported:

- `kWh/100km`
- `mi/kWh`
- `km/kWh`

Preliminary code available at https://github.com/home-assistant/core/compare/dev...jschlyter:home-assistant:energy_distance_units with conversion between any of the three units above.

## Integrations which can benefit from the proposal

Integrations that can benefit are primarily for electric vehicles.

- Audi Connect
- Polestar API
- Porsche Connect
- Skoda Connect
- Volkswagen Connect
- Volvo On Call
- Tessie

## Other work

To be consistent one can also consider a `volume_distance` sensor device class (e.g., `l/km`) for vehicles running on liquid fuel, but that should be addressed separately. Similar to this is `mass_distance` (e.g., `kg/100km`) used for LNG/CNG vehicles.

## Pull Requests

- https://github.com/home-assistant/core/pull/136933
- https://github.com/home-assistant/home-assistant.io/pull/37213
- https://github.com/home-assistant/developers.home-assistant/pull/2548

**Comment by frenck** [comment]
Hi there @jschlyter 👋 

Thanks for opening an arch discussion. 

The proposal makes sense in a way. However, we only want to add support for things like this. There are multiple integrations available in our current core codebase that can leverage this.

Which ones would those be?

../Frenck

**Reply by jschlyter** [reply]
Components for electric vehicles are the primary target, although similar units for combustion cars may be useful (L/100km, mi/ga) as well. Perhaps we want to consider volume per distans if we do energy per distance.

Most of the [car components](https://www.home-assistant.io/integrations/#car) should benefit from this as long as they support electric cars providing these sensors. There are also quite a lot of 3rd party components in HACS that could use this as well. Most EVs are currently supported via HACS, perhaps as their APIs change quite frequently. E.g.,

- Porsche Connect
- Polestar API
- Audi Connect
- Skoda Connect
- Volkswagen Connect

There might be other entities using energy per distance, but I can't think of any directly. Perhaps calories/km or J/km for exercise equipment?

**Reply by AJediIAm** [reply]
Don't forget about Tessie. They already have this in their own app, so I expect it can be added to their home assistant integration as well.

**Comment by jschlyter** [comment]
I suggest we add a more generic support for inverse units while doing this. There's some sample code in https://github.com/jschlyter/home-assistant/blob/9ef4b3c766e2c05596a88b6bc4744926a02614f9/homeassistant/util/unit_conversion.py#L248 and similar for `L/km` and `km/L` should be quite easy based on this.

**Comment by jschlyter** [comment]
Any updates on this issue? Is there anything I can do to move things ahead?

**Reply by frenck** [reply]
> anything I can do to move things ahead?

Have patience.

**Comment by emontnemery** [comment]
@jschlyter we discussed the proposal, while we're positive it's too vague to make a decision on.
 
Please update it so it's clear:
- Exactly what device class(es) should be added
- Exactly what units should be added
- What unit conversions should be supported

This proposal is a good reference: https://github.com/home-assistant/architecture/discussions/1124

**Reply by jschlyter** [reply]
Description updated, including reference to code.

**Reply by gpayer** [reply]
I actually tried to use data from the battery sensor and the "km driven so far" sensor from the Renault integration to calculate the energy consumption per distance for a given point in time. However I'm not sure which helpers to use (probably derivation and then something else combining various helpers), as this does not seem to be explicitly supported.

So introducing these new units should help quite a bit!

**Comment by abmantis** [comment, marked as answer]
The proposal has been approved in the core architectural meeting, but excluding `kWh/100mi`, since that does not seem common. Do we have examples of `kWh/100mi` being used? If not, lets remove it from the proposal so we can move forward.

**Reply by jschlyter** [reply]
I'm happy with removing kWh/100mi, it was included for symmetry – most imperial vehicles use miles/kWh . I'll update my PR shortly.

**Reply by abmantis** [reply]
Can you please also edit this proposal to remove that one?

**Reply by jschlyter** [reply]
Proposal and draft PR updated.

**Comment by nrahbek** [comment]
I think it might be a good idea to also add Wh/km, which is often used in Europe.
